internal/app/machined/pkg/system/services/etcd.go

```diff
@@ -359,3 +317,85 @@ func buildInitialCluster(config runtime.Configurator, name, ip string) (initial
 
 	return initial, nil
 }
+
+func blacklistError(s string) error {
+	return fmt.Errorf("extra etcd arg %q is not allowed", s)
+}
+
+// nolint: gocyclo
+func args(config runtime.Configurator) ([]string, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	metadata, err := metadata.Open()
+	if err != nil {
+		return nil, err
+	}
+
+	ips, err := net.IPAddrs()
+	if err != nil {
+		return nil, fmt.Errorf("failed to discover IP addresses: %w", err)
+	}
+
+	if len(ips) == 0 {
+		return nil, errors.New("failed to discover local IP")
+	}
+
+	a := argumentor.Args{
+		"name":                  hostname,
+		"data-dir":              constants.EtcdDataPath,
+		"listen-peer-urls":      "https://0.0.0.0:2380",
+		"listen-client-urls":    "https://0.0.0.0:2379",
+		"cert-file":             constants.KubernetesEtcdPeerCert,
+		"key-file":              constants.KubernetesEtcdPeerKey,
+		"trusted-ca-file":       constants.KubernetesEtcdCACert,
+		"peer-client-cert-auth": "true",
+		"peer-cert-file":        constants.KubernetesEtcdPeerCert,
+		"peer-trusted-ca-file":  constants.KubernetesEtcdCACert,
+		"peer-key-file":         constants.KubernetesEtcdPeerKey,
+	}
+
+	extraArgs := argumentor.Args(config.Cluster().Etcd().ExtraArgs())
+
+	for k := range a {
+		if extraArgs.Contains(k) {
+			return nil, blacklistError(k)
+		}
+	}
+
+	if extraArgs.Contains("initial-cluster-state") {
+		return nil, blacklistError("initial-cluster-state")
+	}
+
+	a.Set("initial-cluster-state", "new")
```

Comment:
Why not have `"initial-cluster-state": "new"` added to the args list above?

It'd allow us to remove this whole bit because it should get caught via blacklist in the previous check (362-366).

Reply:
Still a WIP